ppa/archive/management/commands/gensim_serialize.py

```diff
@@ -0,0 +1,292 @@
+    strip_multiple_whitespaces, strip_numeric, remove_stopwords, strip_short, stem_text
+
+from django.core.management.base import BaseCommand
+
+from ppa.archive.solr import get_solr_connection
+
+logger = logging.getLogger(__name__)
+
+PREPROCESS_FUNCTIONS = {
+    'lower': lambda x: x.lower(),
+    'strip_tags': strip_tags,
+    'strip_punctuation': strip_punctuation,
+    'strip_multiple_whitespaces': strip_multiple_whitespaces,
+    'strip_numeric': strip_numeric,
+    'remove_stopwords': remove_stopwords,
+    'strip_short': strip_short,
+    'stem_text': stem_text
+}
+
+
+class SolrCorpus:
+    def __init__(self, name, client, collection, doc_id_field='source_id', doc_content_field='content',
+                 page_order_field='order', doc_limit=-1, doc_ids=None, preprocess_fns=None, pbar=True):
+        """
+        A class encapsulating a Solr Client specification, that yields Bag-of-Word vectors on iteration, and thus
+            acts as a Gensim Corpus.
+
+        :param name: A string name of this corpus. Used as a string prefix for generated files.
+        :param client: A SolrClient.SolrClient object used to interface with Solr
+        :param collection: A string representing the Solr collection name.
+        :param doc_id_field: Solr field name for document identifier. Default 'source_id'.
+        :param doc_content_field: Solr field name for document content. Default 'content'.
+        :param page_order_field: Solr field name for page ordering. Default 'order'.
+        :param doc_limit: Max no. of documents to process. The default of -1 means we process ALL documents found.
+        :param doc_ids: List of document identifiers to process. If None, ALL documents found are processed.
+        :param preprocess_fns: A list of single-argument functions to use as preprocessors.
+            See the module gensim.parsing.preprocessing for some typical preprocessing functions.
+        :param pbar: A boolean indicating whether to display a progress bar during corpus generation.
+        """
+        self.name = name
+        self.client = client
+        self.collection = collection
+        self.doc_id_field = doc_id_field
+        self.doc_content_field = doc_content_field
+        self.page_order_field = page_order_field
+        self.doc_limit = doc_limit
+        self.preprocess_fns = preprocess_fns or []
+
+        self.dictionary = Dictionary()
+
+        # doc_id -> dict of k->v mappings
+        # NOTE: We cannot use 'metadata' as GenSim mangles this attribute internally!
+        self._metadata = {}
+
+        # list of strings, populated on first doc retrieval
+        self.metadata_field_names = None
+
+        results = self.client.query(
+            collection=self.collection,
+            query={
+                'q': '*:*',
+                'facet': 'on',
+                'rows': 0,
+                'facet.field': doc_id_field,
+                'facet.limit': self.doc_limit
+            }
```

Comment:
My thinking is that it's perhaps better to just ask Solr for all the information it can provide, both in terms of page counts as well as any other metadata.

An earlier iteration of the program did use the page counts directly from the Django model, but then I thought that's not desirable long-term. The authoritative source of this information would have to be Solr, not Django models, and it would be too tight a coupling to rely on the models anyway. If it wasn't worth storing in the Solr index, then maybe it's not worth retrieving as metadata.

Since getting page counts is a one-time Solr query, I imagine it would have minimal impact on performance.